### PR TITLE
Benchmarking Tooling

### PR DESCRIPTION
**Pull request by araskachoi**
example benchmark run:
in the `./docker/benchmark` directory, run `bash test.sh 5000`
- this will send 5000 transactions (will keep running til 5000 transactions have been sent)
- will result in files that are generated
  - emintcli.log (logs of the output of emintcli process)
  - emintd.log (logs of the output of emintd process) 
  - hashes.json (hashes of all the transactions that were sent)
  - receipts.json (receipt results of the transactions from chain)
  - resource.log (resource consumption logs of emintd and emintcli processes)
- output will give metrics
example output:
```
start time set:  1595299469
end time set:  1595299503
Average CPU Usage [emintd]:  30.83235294117647
Average RAM Usage [emintd]:  5.2823529411764705
Average CPU Usage [emintcli]:  37.126470588235286
Average RAM Usage [emintcli]:  2.667647058823529
1595299499
1595299469
TX per second (TPS):  166.66666666666666
Blocks with Tx:  [0x3 0x4 0x5 0x6 0x7 0x8 0x9]
Block Timestamps:  [1595299469 1595299474 1595299479 1595299484 1595299489 1595299494 1595299499 1595299464]
Transactions:  [801 729 729 724 738 753 526]
Total Transactions:  5000
```

## ~~PLEASE DO NOT  MERGE!~~
~~this PR has a change in the antehandler where it will bypass the exactness (ordering) of the nonce. This is required for sending transactions rapidly.~~
The out-of-sequence nonce checking has been included in https://github.com/ChainSafe/ethermint/pull/550

Closes: #XXX

## Description

______

For contributor use:

- [ ] Targeted PR against correct branch (see [CONTRIBUTING.md](https://github.com/cosmos/cosmos-sdk/blob/master/CONTRIBUTING.md#pr-targeting))
- [ ] Linked to Github issue with discussion and accepted design OR link to spec that describes this work.
- [ ] Code follows the [module structure standards](https://github.com/cosmos/cosmos-sdk/blob/master/docs/building-modules/structure.md).
- [ ] Wrote unit and integration [tests](https://github.com/cosmos/cosmos-sdk/blob/master/CONTRIBUTING.md#testing)
- [ ] Updated relevant documentation (`docs/`) or specification (`x/<module>/spec/`)
- [ ] Added  relevant `godoc` [comments](https://blog.golang.org/godoc-documenting-go-code).
- [ ] Added a relevant changelog entry to the `Unreleased` section in `CHANGELOG.md`
- [ ] Re-reviewed `Files changed` in the Github PR explorer

______

For admin use:

- [ ] Added appropriate labels to PR (ex. `WIP`, `R4R`, `docs`, etc)
- [ ] Reviewers assigned
- [ ] Squashed all commits, uses message "Merge pull request #XYZ: [title]" ([coding standards](https://github.com/tendermint/coding/blob/master/README.md#merging-a-pr))

### CODE REVIEW

File: docker/Dockerfile [hunk 1]
```diff
@@ -0,0 +1,13 @@
+FROM golang:1.13
```
Comment (fedekunze) [inline docker/Dockerfile]:
```suggestion
FROM golang:1.14
```

File: app/ante/eth.go [hunk 2]
```diff
@@ -237,7 +237,7 @@ func (nvd NonceVerificationDecorator) AnteHandle(ctx sdk.Context, tx sdk.Tx, sim
 	}
 
 	seq := acc.GetSequence()
-	if msgEthTx.Data.AccountNonce != seq {
+	if msgEthTx.Data.AccountNonce < seq {
```
Comment (fedekunze) [inline app/ante/eth.go]: why the change?

File: docker/benchmarking/benchmark/geth-genesis.go [hunk 3]
```diff
@@ -0,0 +1,121 @@
...
+import (
+	"encoding/json"
+	"io/ioutil"
+	"os"
+	"strconv"
+
+	"github.com/urfave/cli"
+)
+
+type Genesis struct {
```
Comment (fedekunze) [inline docker/benchmarking/benchmark/geth-genesis.go:13]: nit add godoc comment

File: docker/benchmarking/benchmark/geth-genesis.go [hunk 4]
```diff
@@ -0,0 +1,121 @@
...
+	Config struct {
```
Comment (fedekunze) [inline docker/benchmarking/benchmark/geth-genesis.go]: move out of the struct and pass it as a field

File: docker/benchmarking/benchmark/geth-genesis.go [hunk 5]
```diff
@@ -0,0 +1,121 @@
...
+		HomesteadBlock      int64       `json:"homesteadBlock"`
+		ByzantiumBlock      int64       `json:"byzantiumBlock"`
+		ConstantinopleBlock int64       `json:"constantinopleBlock"`
+		PetersburgBlock     int64       `json:"petersburgBlock"`
+		Consensus           interface{} `json:"clique"`
+	} `json:"config"`
+	Difficulty string      `json:"difficulty"`
+	GasLimit   string      `json:"gasLimit"`
+	ExtraData  string      `json:"extraData"`
+	Alloc      interface{} `json:"alloc"`
```
Comment (fedekunze) [inline docker/benchmarking/benchmark/geth-genesis.go:18]: ditto

File: docker/benchmarking/benchmark/geth-genesis.go [hunk 6]
```diff
@@ -0,0 +1,121 @@
...
+}
+type Balance struct {
```
Comment (fedekunze) [inline docker/benchmarking/benchmark/geth-genesis.go:22]: ditto godoc

File: docker/benchmarking/benchmark/geth-genesis.go [hunk 7]
```diff
@@ -0,0 +1,121 @@
...
+	if err != nil {
+		jsonFile, err = os.Open("templ-genesis.json")
+		if err != nil {
+			return Genesis{}, err
+		}
+	}
+	defer jsonFile.Close()
+	byteValue, _ := ioutil.ReadAll(jsonFile)
+	var genesis Genesis
+	json.Unmarshal(byteValue, &genesis)
```
Comment (fedekunze) [inline docker/benchmarking/benchmark/geth-genesis.go]: check error

File: docker/benchmarking/benchmark/geth-genesis.go [hunk 8]
```diff
@@ -0,0 +1,121 @@
...
+	return genesis, nil
+}
+
+func writeGenesis(data Genesis) error {
+	file, err := json.MarshalIndent(data, " ", " ")
+	if err != nil {
+		return err
+	}
+
+	err = ioutil.WriteFile("genesis.json", file, 0644)
```
Comment (fedekunze) [inline docker/benchmarking/benchmark/geth-genesis.go]: address lint issue

File: docker/benchmarking/benchmark/tx.go [hunk 9]
```diff
@@ -0,0 +1,589 @@
...
+	"strconv"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/ethereum/go-ethereum/common/hexutil"
+	"github.com/urfave/cli"
+)
+
+type Request struct {
```
Comment (fedekunze) [inline docker/benchmarking/benchmark/tx.go:23]: ditto godoc comments in this file

**Review by fedekunze — COMMENTED** [review]
Looks good! Will test tomorrow morning. Pending lint fixes

File: docker/benchmarking/emint/init.sh [hunk 10]
```diff
@@ -0,0 +1,59 @@
...
+
+# Allocate genesis accounts (cosmos formatted addresses)
+ethermintd add-genesis-account $(ethermintcli keys show $FKEY -a) 1000000000000000000photon,1000000000000000000stake
+
+# Generate 30 accounts
+echo -e 'Generating Accounts\n'
+for i in $(seq 1 $ACCTS)
+do
+  ethermintcli keys add $TESTKEY$i --algo "eth_secp256k1"
+  ethermintd add-genesis-account $(ethermintcli keys show $TESTKEY$i -a) 1000000000000000000photon,1000000000000000000stake
```
Comment (fedekunze) [inline docker/benchmarking/emint/init.sh]: these need to be aphoton now

File: docker/README.md [hunk 11]
```diff
@@ -0,0 +1,8 @@
+# Docker Benchmarking
+
+## Getting Started
+
+Run the command in the root directory of this project
+```
+$ docker build -f ./docker/Dockerfile .
```
Comment (fedekunze) [inline docker/README.md]:
```suggestion

```bash
docker build -f ./docker/Dockerfile . 
```

**Review by fedekunze — COMMENTED** [review]
thanks @araskachoi do you think we can address the final comments so that we can merge this PR?

File: app/ante/eth.go (hunk 2 above)
Comment (araskachoi) [reply app/ante/eth.go]: the sequence goes out of order as it takes some time for an account nonce to either be updated or reverted. thus, we need this in order to bypass the nonce for these benchmarking tests. Any tips on how to work around this? I hesitate to merge this branch because of this change

Comment (fedekunze) [reply app/ante/eth.go]: I see. Then I'd probably update the docker README and suggest to change this line if someone wants to reproduce the benchmarks

Comment (araskachoi) [reply app/ante/eth.go]: https://github.com/ChainSafe/ethermint/pull/550

**Comment by fedekunze** [comment]
@araskachoi should we reopen this PR? feel free to do so and ping me once it's R4R 👍
